### PR TITLE
URI versioning added to server.

### PR DESCRIPTION
**Pull request by kshakir**
Patched swagger-ui routes and added tests.

### CODE REVIEW

**Comment by kshakir** [comment]
Picking @mcovarr as first reviewer. Fine with merge going in sooner, or later after incorporating other pending server changes.

File: src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 1]
```diff
@@ -22,54 +19,40 @@ object SwaggerService {
...
-    }
+  def possibleRoutes = workflowRoutes ~ swaggerService.routes ~ swaggerService.uiRoutes
 
   def receive = runRoute(possibleRoutes)
 }
 
+object CromwellApiService {
+  /**
+   * Comma separated list of optional endpoint versions. The last value is the latest version.
+   */
```
Comment (mcovarr) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: maybe a `List` with the current value at the head?

File: build.sbt [hunk 2]
```diff
@@ -15,7 +15,7 @@ val googleClientApiV = "1.20.0"
 
 libraryDependencies ++= Seq(
   "com.gettyimages" %% "spray-swagger" % "0.5.1",
-  "org.webjars" % "swagger-ui" % "2.0.24",
+  "org.webjars" % "swagger-ui" % "2.1.8-M1",
```
Comment (mcovarr) [inline build.sbt]: does this need to be a milestone release?

File: src/main/scala/cromwell/webservice/SwaggerConfigHttpService.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    else
+      default
+  }
+}
+
+trait SwaggerUiHttpService extends HttpService {
+  this: SwaggerHttpService =>
+
+  /**
+   * @return path to serve the swagger-uu
```
Comment (mcovarr) [inline src/main/scala/cromwell/webservice/SwaggerConfigHttpService.scala]: do uu mean ui?

File: src/test/scala/cromwell/webservice/SwaggerConfigHttpServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import spray.http.StatusCodes
+import spray.testkit.ScalatestRouteTest
+
+class SwaggerConfigHttpServiceSpec extends FlatSpec with SwaggerConfigHttpService with ScalatestRouteTest with Matchers {
+  override def actorRefFactory = system
+
+  override def apiTypes = Nil
+
+  // Version values in application.conf, possibly reference.conf,
+  // and this test conf must be manually synced with build.sbt
```
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/SwaggerConfigHttpServiceSpec.scala]: yeah that's kind of a bummer... :frowning:

**Comment by mcovarr** [comment]
:+1: from me with nits, assigning to @geoffjentry

File: src/main/scala/cromwell/webservice/CromwellApiService.scala (hunk 1 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: Part of me wants to say to make a case class with a current version and previous versions but that sounds an awful lot like a list :)

File: src/main/scala/cromwell/webservice/SwaggerConfigHttpService.scala [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  // Validated by SwaggerConfigHttpServiceSpec
+  override lazy val apiVersion: String = swaggerConfig.getString("api.version")
+  override lazy val swaggerUiVersion: String = swaggerConfig.getString("ui.version")
+
+  override lazy val docsPath: String = optString("docsPath", super.docsPath)
+  override lazy val swaggerVersion: String = optString("swaggerVersion", super.swaggerVersion)
+  override lazy val uiPath: String = optString("ui.path", super.uiPath)
+
+  override lazy val apiInfo: Option[ApiInfo] =
+    if (!swaggerConfig.hasPath("api.info"))
```
Comment (geoffjentry) [inline src/main/scala/cromwell/webservice/SwaggerConfigHttpService.scala]: Could the branches be reversed so it's `if (swaggerConfig.hasPath....)`?

**Comment by geoffjentry** [comment]
:+1: once comments have been addressed

File: src/main/scala/cromwell/webservice/CromwellApiService.scala (hunk 1 above)
Comment (kshakir) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: Unless I'm being dense, as this value is used in an annotation, and must stay a static / non-generated string? Updated this variable name and comment to hopefully make it clearer where it's used below.

File: build.sbt (hunk 2 above)
Comment (kshakir) [inline build.sbt]: Apparently a "stable" release has been released _after_ 2.1.8-M1 that I missed, due to the way versions are sorted. "Updated" to 2.1.1.

**Comment by kshakir** [comment]
Biggest change for re-review: While breaking the URIs with versions in this PR, I just pushed another commit that also changed the endpoints to be consistently plural across all REST endpoints.

File: src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 6]
```diff
@@ -134,15 +139,16 @@ trait CromwellApiService extends HttpService with PerRequestCreator  {
...
     produces = "application/json"
-    //,response = classOf[WorkflowOutputs]
   )
   @ApiImplicitParams(Array(
-    new ApiImplicitParam(name = "id", required = true, dataType = "string", paramType = "path", value = "workflow")
+    new ApiImplicitParam(name = "version", required = true, dataType = "string", paramType = "path", value = "API Version", allowableValues = CromwellApiService.VersionAllowableValues),
+    new ApiImplicitParam(name = "id", required = true, dataType = "string", paramType = "path", value = "Workflow Identifier")
   ))
   @ApiResponses(Array(
     new ApiResponse(code = 200, message = "Successful Request"),
```
Comment (mcovarr) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: Can the `response` attribute that was removed from the `@ApiOperation` be relocated to the appropriate `@ApiResponse` here as it was in other APIs?  If I recall Kristian was having compilation problems around this annotation.  :frowning:

**Comment by mcovarr** [comment]
No strong opinions on the `workflow` vs `workflows` naming of endpoints.  I believe our API began life as a copy of Rawls, but currently Rawls seems consistent with how you've got this now.

**Comment by geoffjentry** [comment]
Still :+1: from me

File: src/main/scala/cromwell/webservice/CromwellApiService.scala (hunk 6 above)
Comment (kshakir) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: Added a more explicit comment around this line.

**Pull request merged**